_SPIKE_.md

```diff
@@ -0,0 +1,163 @@
+
+I feel it's important to use model-named classes (e.g. `ProductController`)
+instead of reusing the same controller/repository class name for all models.
+Model-named classes make error troubleshooting and debugging easier, because
+stack traces include model name in function names. (Compare
+`ProductController.replaceById` with `CrudRestControllerImpl.replaceById` -
+which one is more useful?)
+
+In my proposal, model-config files are in JSON format to make programmatic edits
+easier. This has a downside in TypeScript projects - these config files must
+live outside `src` because TypeScript does not copy arbitrary JSON files.
+
+### Extensibility & customization options
+
+The proposed design enables the following opportunities to extend and customize
+the default behavior of API endpoints:
+
+- App developers can create & bind a custom repository class, this allows them
+  e.g. to implement functionality similar to LB3 Operation Hooks.
+
+- App developers can implement their own api-builder plugins, replacing the
+  repository & controller builders provided by LB4 by their own logic.
+
+- Model configuration schema is extensible, individual plugins can define
+  additional model-config options to further tweak the behavior of API
+  endpoints.
+
+## How to review the spike
+
+I have updated `examples/todo` application to leverage this new functionality,
+you can look around the updated code to see what the user experience will look
+like. Take a look at how acceptance level tests are accessing the repository
+created by `rest-crud` plugin.
+
+The booter is implemented in
+[`packages/booter-rest/src/rest.booter.ts`](https://github.com/strongloop/loopback-next/blob/spike/crud-rest-booter/packages/booter-rest/src/rest.booter.ts).
+Initially, I created a new package for this booter, but on the second thought, I
+think this booter is generic enough to be included in `@loopback/boot` under a
+different name `ModelApiBooter`.
+
+The plugin (extension) contract is defined in two files:
+
+- [`packages/model-api-builder/src/model-api-builder.ts`](https://github.com/strongloop/loopback-next/blob/spike/crud-rest-booter/packages/model-api-builder/src/model-api-builder.ts)
+- [`loopback-next/packages/model-api-builder/src/model-api-config.ts`](https://github.com/strongloop/loopback-next/blob/spike/crud-rest-booter/packages/model-api-builder/src/model-api-config.ts)
+
+The CRUD REST API builder:
+[`packages/rest-crud/src/crud-rest-builder.plugin.ts`](https://github.com/strongloop/loopback-next/blob/spike/crud-rest-booter/packages/rest-crud/src/crud-rest-builder.plugin.ts)
+
+The remaining changes are small tweaks & improvements of existing packages to
+better support this spike.
+
+## Open questions:
+
+- Where to keep model config files?
+
+  - `/public-models/product.config.json` (JSON, must be outside src)
+  - `/src/public-models/product-config.ts` (TS, can be inside src, more
+    flexible)
+
```

Comment:
We can follow the same convention as `datasources`:

- src/resources/my-model.resource.json
- src/resources/my-model.resource.ts

We can discuss what's the best name for `resources`.

Reply:
The approach we are using for datasources is actually problematic, see https://github.com/strongloop/loopback-next/issues/2794

> Currently the `lb4 datasource` command will create a `{name}.datasource.ts` file and a `{name}.datasource.json` file. Since these files have the same base name, it causes issues with imports while running with `ts-node` (useful for development). By default, `ts-node` will import the JSON file over the TS file, causing datasource construction to fail. For more info on how `ts-node` works, see https://github.com/TypeStrong/ts-node/issues/769
>
> I was able to workaround this issue by manually renaming the json file, but it would be good if the datasource generator didn't create two files with the same base name.

Also datasources are different from model configs. A datasource file exports a datasource class, therefore we need a TS file. The configuration json file is kind of optional for datasources.

Model configs are exporting a config object only, therefore we have the option to decide whether to use JSON or TS.

As I was thinking about this more, I am leaning towards using TS files.
- It's more consistent with the approach we have for other artifacts, most notably model definitions.
- It gives a lot of flexibility to application developers, they can e.g. parse env variables to decide certain aspects of model configuration.

Reply:
Unfortunately, the convention used by datasources does not work well.

Quoting from https://github.com/strongloop/loopback-next/issues/2794:
> Currently the `lb4 datasource` command will create a `{name}.datasource.ts` file and a `{name}.datasource.json` file. Since these files have the same base name, it causes issues with imports while running with `ts-node` (useful for development). By default, `ts-node` will import the JSON file over the TS file, causing datasource construction to fail.

In my view, datasources are different from model-config files.
- For datasources, we primarily need the `.ts` file that is exporting the datasource class. The `.json` file with the configuration is just an optional support.
- For model config files, there is no class to export, the booter needs a data object with configuration options.

As I was thinking about this more, I am leaning towards using TS files in `src` folder. I feel this is more consistent with the approach we use for all other artifacts (models, repositories, etc.). It also enables application developers to conditionally customize model config, e.g. depending on `process.env` variables.

I am going to update this part of the spike.